### PR TITLE
Issue #403: Expose attributes and adjustments to Views

### LINKED ISSUE #403: Expose attributes to views
Like written here: https://www.drupal.org/project/ubercart/issues/895236 - having the ability to use attributes within views (e.g. to export a range of orders as csv file with attributes as separate fields) would be great.

I don't know whether the D7 patch works and could be ported.

### PR DESCRIPTION
**Pull request by bugfolder**
Fixes https://github.com/backdrop-contrib/ubercart/issues/403.

### CODE REVIEW

File: uc_attribute/uc_attribute.module [hunk 1]
```diff
@@ -212,6 +212,16 @@ function uc_attribute_permission() {
   );
 }
 
+/**
+ * Implements hook_views_api().
+ */
+function uc_attribute_views_api() {
+  return array(
+    'api' => '2.0',
+    'path' => drupal_get_path('module', 'uc_attribute') . '/views',
```
Comment (argiepiano) [inline uc_attribute/uc_attribute.module]: This should be backdrop_get_path

File: uc_attribute/views/uc_attribute.views.inc [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+      'help' => 'Filter by this attribute if it has its own SKU on the products\'s adjustments.',
+      'filter' => array(
+        'handler' => 'UcAttributeHandlerFilterProductAttr',
+      ),
+    );
+  }
+
+  $data['uc_product_adjustments']['table']['group'] = t('Product adjustments');
+
+  $uc_product_adjustments = drupal_get_schema_unprocessed('uc_attribute', 'uc_product_adjustments');
```
Comment (argiepiano) [inline uc_attribute/views/uc_attribute.views.inc]: This should be backdrop_get_schema_unprocessed

File: uc_attribute/views/uc_attribute_handler_field_order_product_attribute.inc [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    return $options;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function render($values) {
+    $data = unserialize($values->{$this->field_alias});
+    $result = "";
+    if (is_array($data['attributes'])) {
```
Comment (argiepiano) [inline uc_attribute/views/uc_attribute_handler_field_order_product_attribute.inc]: This throws `Notice: Undefined index: attributes in UcAttributeHandlerFieldOrderProductAttribute->render() (line 25 ` when you have some ordered products without attributes.

Suggests:

```php
if (!empty($data['attributes'])) {
```

I believe `$data['attributes']` will always be an array if it's not empty.

File: uc_attribute/uc_attribute.module (hunk 1 above)
Comment (bugfolder) [reply uc_attribute/uc_attribute.module]: Yes, fixed, thanks!

File: uc_attribute/views/uc_attribute.views.inc (hunk 2 above)
Comment (bugfolder) [reply uc_attribute/views/uc_attribute.views.inc]: Yes, fixed, thanks!

File: uc_attribute/views/uc_attribute_handler_field_order_product_attribute.inc (hunk 3 above)
Comment (bugfolder) [reply uc_attribute/views/uc_attribute_handler_field_order_product_attribute.inc]: Yes, fixed, thanks!

**Comment by bugfolder** [comment]
Closing; issue fixed by the creation of the contrib Ubercart Attribute Views module.

**Pull request closed without merging**